### ISSUE TITLE
Add abstraction for authentication strategy

### ISSUE THREAD

**Issue by jannyHou** [strongloop/loopback-next#2466]
## Feature

Define an interface to describe the authentication strategy. It should contain

- [x] a function `authenticate` that verifies the request.
- [ ] ~function `authenticate` should take in an `option` to perform the auth action accordingly.~
- [x] document if people want to use services in the authentication strategy, they can inject the service in the strategy constructor 

**Acceptance Criteria**
- [x] Verify the interface works by refactoring the Shopping Cart example to utilize it.
- [x] Verify the interface works by refactoring the Shopping Cart example to utilize it and strategies registered via extension points.

Parent story see https://github.com/strongloop/loopback-next/issues/1035#issuecomment-465311972 (step 1b)

**Comment by bajtos** [comment]
How is this story different from #2311 and #2312? Is it a pre-requisite that must be implemented first, before we can work on an extension point?

As I wrote in https://github.com/strongloop/loopback-next/issues/2312#issuecomment-469598086, I think the authentication strategy abstraction should be defined around what LoopBack-specific strategies need and allow us to write an adapter to convert Passport strategies into LoopBack strategies. 

What I am trying to say: The approach shown in the current version of `@loopback/authentication`, where we use Passport's `Strategy` interface as the API contract between LoopBack strategies and the authentication component, is something we should abandon in favor of a more flexible solution.

**Comment by jannyHou** [comment]
@bajtos 
> How is this story different from #2311 and #2312? Is it a pre-requisite that must be implemented first, before we can work on an extension point?

This story simply creates the interface for auth strategy, the interface will be mostly the same as what is proposed in the [design file](https://github.com/strongloop/loopback-next/pull/2576/files#diff-d7368ea1b2895005672ca0963f8ef9e9R6)

**Comment by samarpanB** [comment]
@bajtos we have done oauth 2 integration as a separate component in our LB4 application very recently. We followed the guidelines shared [here](https://github.com/strongloop/loopback-next/blob/master/packages/authentication/README.md). I am not too sure what different aspect is getting implemented here or any of the authentication tasks. But **if there is any way I can help/contribute, do let me know**. I can share our implementation as well, if needed. That's not so robust but it was workable for us. We also did authorisation component but I would mention that in the appropriate issue.
